src/hotspot/cpu/s390/gc/shared/barrierSetNMethod_s390.cpp

```diff
@@ -23,18 +23,103 @@
  */
 
 #include "precompiled.hpp"
+#include "asm/assembler.inline.hpp"
+#include "code/codeBlob.hpp"
+#include "code/nativeInst.hpp"
+#include "code/nmethod.hpp"
 #include "gc/shared/barrierSetNMethod.hpp"
 #include "utilities/debug.hpp"
 
+class NativeMethodBarrier: public NativeInstruction {
+  private:
+    static const int GUARD_INSTRUCTION_OFFSET = 3*6; // bytes
+
+    address get_barrier_start_address() const {
+      return NativeInstruction::addr_at(0);
+    }
+
+    NativeMovRegMem* get_patchable_instruction_handle() const {
+      address guard_addr = get_barrier_start_address() + GUARD_INSTRUCTION_OFFSET;
+
+      return reinterpret_cast<NativeMovRegMem*>(guard_addr);
+    }
+
+  public:
+    static const int BARRIER_TOTAL_LENGTH = GUARD_INSTRUCTION_OFFSET + 2*6 + 2; // bytes
+
+    int get_guard_value() const {
+      NativeMovRegMem* guard_addr = get_patchable_instruction_handle();
+
+      // Return guard instruction and value
+      return guard_addr->offset();
+    }
+
+    void set_guard_value(int value) {
+      NativeMovRegMem* guard_addr = get_patchable_instruction_handle();
+
+      // Set guard instruction and value
+      guard_addr->set_offset(value);
+    }
+
+    void verify() const {
+      int offset = 0;
+      const address start = get_barrier_start_address();
+
+      assert(Assembler::is_equal(start[offset],  IIHF_ZOPC, RIL_MASK), "check load_const (1/2)");
+      offset += Assembler::instr_len(IIHF_ZOPC);
+      assert(Assembler::is_equal(start[offset],  IILF_ZOPC, RIL_MASK), "check load_const (2/2)");
+      offset += Assembler::instr_len(IILF_ZOPC);
+      assert(Assembler::is_equal(start[offset], LG_ZOPC, RIL_MASK), "check LG");
+      offset += Assembler::instr_len(LG_ZOPC);
+      assert(Assembler::is_equal(start[offset], CFI_ZOPC, RIL_MASK), "check CFI");
+      get_patchable_instruction_handle()->verify();
+      offset += Assembler::instr_len(CFI_ZOPC);
+      assert(Assembler::is_equal(start[offset], LARL_ZOPC, RIL_MASK), "check LARL");
+      offset += Assembler::instr_len(LARL_ZOPC);
+      assert(Assembler::is_equal(start[offset], BCR_ZOPC, RIL_MASK), "check BCR");
+      offset += Assembler::instr_len((unsigned long)BCR_ZOPC);
+
```

Comment:
I don't like specifying the same information (here: instruction opcode) multiple times. I would suggest you increment the offset with code like
`offset += Assembler::instr_len(&start[offset]);`

Reply:
I like this. I have made this change.